lib/SILOptimizer/Mandatory/SILGenCleanup.cpp

```diff
@@ -104,39 +107,189 @@ struct SILGenCleanup : SILModuleTransform {
   void run() override;
 
   bool completeOSSALifetimes(SILFunction *function);
+  template <typename Range>
+  bool completeLifetimesInRange(Range const &range,
+                                OSSALifetimeCompletion &completion,
+                                BasicBlockSet &completed);
 };
 
+/// Progress made so far on a walk beginning at an unreachable-terminated block.
+struct Walk {
+  SILBasicBlock *current;
+  SmallPtrSet<SILBasicBlock *, 16> seen;
+};
+
+/// Populate `roots` with the last blocks that are discovered via backwards
+/// walks along any non-repeating paths starting at the ends in `walks`.
+void collectReachableRoots(SILFunction *function, SmallVectorImpl<Walk> &walks,
```

Comment:
This algorithm looks a bit complicated. Would this also work?

1. add blocks which are reachable from `unreachable` in BasicBlockSet _s_ (by going backwards in the CFG)
2. For each block: if the block is unreachable _and_ has no predecessors _and_ is in _s_: add it to the roots
3. add blocks which are reachable from roots in BasicBlockSet _t_ (by going forward in the CFG)
4. pick one unreachable block which is not in _t_ and add it to the root (to handle unreachable loops)
5. continue with 3. until all unreachable blocks are in _t_.

Reply:
This sounds better, thanks!  Ill give it a shot.

Reply:
Done except that a block in `s` but not `t` is not necessarily in an unreachable loop, only forwards-reachable from (at least) one, so it's necessary to walk backwards from such a block to find a representative block of the loop (see `unreachable_def_4`).  Also, the initial backward walk begins at all dead-end blocks; I was previously only starting at unreachable-terminated blocks which misses unreachable infinite loops.